### PR TITLE
[zero] add test sharded optim with cpu adam

### PR DESCRIPTION
**Pull request by ver217**
1. Test `ShardedOptimV2` with `CPUAdam`. Grad and fp32 master param are offloaded to CPU.

### CODE REVIEW

File: tests/test_zero_data_parallel/test_sharded_optim_v2_with_cpu_adam.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        model, train_dataloader, test_dataloader, optimizer, criterion = get_components_func()
+        model = model(checkpoint=True).cuda()
+        zero_model = ShardedModelV2(copy.deepcopy(model), shard_strategy, offload_config={'device': 'cpu'})
+        if dist.get_world_size() > 1:
+            model = DDP(model)
+        optim = Adam(model.parameters(), lr=1e-3)
+        sharded_optim = ShardedOptimizerV2(CPUAdam(zero_model.parameters(), lr=1e-3),
+                                           zero_model,
+                                           shard_strategy,
+                                           initial_scale=2**5,
```
Comment (feifeibear) [inline tests/test_zero_data_parallel/test_sharded_optim_v2_with_cpu_adam.py:65]: initial_scale setting is consist with the `Adam(model.parameters(), lr=1e-3)`?

**Review by FrankLeeeee — APPROVED** [review]

File: tests/test_zero_data_parallel/test_sharded_optim_v2_with_cpu_adam.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+def run_dist(rank, world_size, port):
+    colossalai.launch(config=CONFIG, rank=rank, world_size=world_size, host='localhost', port=port, backend='nccl')
+    test_models = ['repeated_computed_layers', 'resnet18', 'bert']
+    for model_name in test_models:
+        get_components_func = non_distributed_component_funcs.get_callable(model_name)
+        shard_strategy = TensorShardStrategy()
+        model, train_dataloader, test_dataloader, optimizer, criterion = get_components_func()
+        model = model(checkpoint=True).cuda()
+        zero_model = ShardedModelV2(copy.deepcopy(model), shard_strategy, offload_config={'device': 'cpu'})
+        if dist.get_world_size() > 1:
```
Comment (feifeibear) [inline tests/test_zero_data_parallel/test_sharded_optim_v2_with_cpu_adam.py:59]: can we still use DDP when world size = 1

File: tests/test_zero_data_parallel/test_sharded_optim_v2_with_cpu_adam.py (hunk 1 above)
Comment (ver217) [reply tests/test_zero_data_parallel/test_sharded_optim_v2_with_cpu_adam.py:65]: Adam doesn't have grad scaler.

Comment (FrankLeeeee) [reply tests/test_zero_data_parallel/test_sharded_optim_v2_with_cpu_adam.py:65]: it should be set to initial_scale = 1

Comment (FrankLeeeee) [reply tests/test_zero_data_parallel/test_sharded_optim_v2_with_cpu_adam.py:65]: but since grad is not checked, I think this is fine as well?

Comment (feifeibear) [reply tests/test_zero_data_parallel/test_sharded_optim_v2_with_cpu_adam.py:65]: I see

File: tests/test_zero_data_parallel/test_sharded_optim_v2_with_cpu_adam.py (hunk 2 above)
Comment (ver217) [reply tests/test_zero_data_parallel/test_sharded_optim_v2_with_cpu_adam.py:59]: Yes, we can. But I think it's meaningless.

**Pull request merged**